File: lecture_014_Tree/treePrac.java
import java.util.*;
public class treePrac{
    
    public static class Node{
        int data;
        Node left;
        Node right;
    
        Node(){
    
        }
    
        Node(int data){
            this.data = data;
        }
    
        Node(int data,Node left,Node right){
            this.data = data;
            this.left = left;
            this.right = right;
    
        }        
    }

    public static void display(Node node) {
        if(node==null){
            return;
        }
        String ans="";
        ans+=node.left!=null?node.left.data:".";
        ans+="<-"+node.data+"->";
        ans+=node.right!=null?node.right.data:".";

        System.out.println(ans);
        display(node.left);
        display(node.right);
    }


    public static Node createBST(int []arr,int si,int ei){
        if(si>ei) return null;
        int mid = si+(ei-si)/2;
        Node node = new Node(arr[mid]);
        node.left = createBST(arr,si,mid-1);
        node.right = createBST(arr,mid+1,ei);
        return node;
    }

    //recursive
    public static boolean findInBST(Node node,int data){
        if(node==null) return false;
        if(node.data == data) return true;
        boolean res=false;
        if(data<node.data) res=res || findInBST(node.left,data);
        else res=res || findInBST(node.right,data);
        return res;
    }

    public static boolean findInBSTitr(Node node,int data){
        if(node==null) return false;
        Node temp = node;

        while(temp!=null){
            if(temp.data==data) return true;
            else if(data<temp.data) temp=temp.left;
            else temp=temp.right;
        }
        return false;
    }

    static int prev = (int)-1e7;
    public static boolean isBST(Node node){
        if(node==null) return true;
        boolean res = true;
        res = res && isBST(node.left);

        if(prev>node.data) return false;
        prev=node.data;

        res = res && isBST(node.right);
        return res;

    }

    public static int  maxEle(Node node){
        Node tmp=node;
        while(tmp.right!=null) tmp=tmp.right;
        return tmp.data;
    }
    public static int minEle(Node node){
        Node tmp=node;
        while(tmp.left!=null) tmp=tmp.left;
        return tmp.data;
    }
    
    public static int LCAofBST(Node node,int p,int q){
        Node temp = node;
        while(temp!=null){
            if(temp.data>p && temp.data>q) temp=temp.left;
            else if(temp.data<p && temp.data<q) temp=temp.right;
            else return temp.data;
        }
        return -1;
    }

    //kfar
    public static boolean nodetorootpath(Node node,int data,ArrayList<Node>list){
        if(node==null) return false;
        boolean res=false;
        if(node.data == data) {
            list.add(node);
            return true;
        }
        res=res || nodetorootpath(node.left,data,list);
        res=res || nodetorootpath(node.right,data,list);
        if(res) list.add(node);
        return res;
    }

    public static void kdown(Node node,int level,Node avoid){
        if(node==avoid || node==null) return;

        if(level==0) System.out.print(node.data+ " ");

        kdown(node.left,level-1,avoid);
        kdown(node.right,level-1,avoid);
    }

    public static void kfar(int data,Node node,int k){
        ArrayList<Node> list = new ArrayList<>();
        nodetorootpath(node,data,list);
        Node avoid=null; 
        for(int i=0;i<list.size();i++){
            kdown(list.get(i),k-i,avoid);
            avoid=list.get(i);
        }
    }

    //view set
    public static void lineWiseLevelOrder(Node node){
        LinkedList<Node> list = new LinkedList<>();
        list.add(node);
        int level=0;
        
        while(list.size()!=0){
            int size=list.size();
            System.out.print("Level " + level + " -> ");
            level++;
            while(size-->0){
                Node rn = list.removeFirst();
                System.out.print(rn.data+", ");
                if(rn.left!=null) list.addLast(rn.left);
                if(rn.right!=null) list.addLast(rn.right);
            }
            System.out.println(); 
        }
    }

    public static void leftView(Node node){
        LinkedList<Node> list = new LinkedList<>();
        list.add(node);
        int level=0;
        
        while(list.size()!=0){
            int size=list.size();
            System.out.print("Level " + level + " -> ");
            level++;
            Node l = list.getFirst();  //or peekfirst()
            System.out.print(l.data);
            while(size-->0){
                Node rn = list.removeFirst(); 
                //System.out.print(rn.data+", ");
                if(rn.left!=null) list.addLast(rn.left);
                if(rn.right!=null) list.addLast(rn.right);
            }
            System.out.println(); 
        }
    }

    public static void rightView(Node node){
        LinkedList<Node> list = new LinkedList<>();
        list.add(node);
        int level=0;
        
        while(list.size()!=0){
            int size=list.size();
            System.out.print("Level " + level + " -> ");
            level++;
            Node l = list.peekLast();  // or getLast()
            System.out.print(l.data);
            while(size-->0){
                Node rn = list.removeFirst();
                //System.out.print(rn.data+", ");
                if(rn.left!=null) list.addLast(rn.left);
                if(rn.right!=null) list.addLast(rn.right);
            }
            System.out.println(); 
        }
    }

//vertical order

    static int leftMinValue=0;
    static int rightMaxValue=0;

    public static void width(Node node,int lev){
        if(node==null) return;
 
        leftMinValue=Math.min(leftMinValue,lev);
        rightMaxValue=Math.max(rightMaxValue,lev);
        
        width(node.left, lev - 1);
        width(node.right, lev + 1);
    } 

    public static class pairVO{
        Node node;  //actual Node
        int vl=0;  // vertical Level
        public pairVO(Node node,int vl){
            this.node=node;
            this.vl=vl;
        }
    }
 
    public static void verticalOrder(Node node){
        width(node,0);
        int n=rightMaxValue - leftMinValue + 1;
        ArrayList<ArrayList<Integer>> ans=new ArrayList<>(); // vector<vector<int>> (n,vector<int>());
        for(int i=0;i<n;i++)
          ans.add(new ArrayList<>());
       
 
        LinkedList<pairVO> que=new LinkedList<>();
        que.addLast(new pairVO(node,-leftMinValue));
 
        while(que.size()!=0){
            int size=que.size();
            while(size--> 0){
                pairVO rpair=que.removeFirst();
                ans.get(rpair.vl).add(rpair.node.data);
                if(rpair.node.left!=null) que.addLast(new pairVO(rpair.node.left, rpair.vl - 1));
                if(rpair.node.right!=null) que.addLast(new pairVO(rpair.node.right,rpair.vl + 1));    
            }
        }
 
        for(ArrayList<Integer> ar: ans)
           System.out.println(ar);
        System.out.println();
    }

    public static void verticalOrderSum(Node node){
        width(node,0);
        int n=rightMaxValue - leftMinValue + 1;
        int [] ans = new int[n];
 
        LinkedList<pairVO> que=new LinkedList<>();
        que.addLast(new pairVO(node,-leftMinValue));
 
        while(que.size()!=0){
            int size=que.size();
            while(size--> 0){
                pairVO rpair=que.removeFirst();
                ans[rpair.vl] += rpair.node.data;
                if(rpair.node.left!=null) que.addLast(new pairVO(rpair.node.left, rpair.vl - 1));
                if(rpair.node.right!=null) que.addLast(new pairVO(rpair.node.right,rpair.vl + 1));    
            }
        }
 
        for(int ele: ans)
           System.out.println(ele);
        System.out.println();
    }

    static int leftMinDiag = 0;
    public static void widthDiag(Node node,int level){
        if(node==null) return;
        leftMinDiag = Math.min(leftMinDiag,level);
        widthDiag(node.left,level-1);
        widthDiag(node.right,level+0);
    }

    public static void diagonalOrder(Node node){
        widthDiag(node,0);
        int n = -leftMinDiag + 1;
        ArrayList<ArrayList<Integer>> ans = new ArrayList<>();
        // for(ArrayList<Integer> ar:ans){
        //     ar=new ArrayList<>();
        // }
        for(int i=0;i<n;i++){
            ans.add(new ArrayList<>());
        }
        LinkedList<pairVO> que = new LinkedList<>();
        que.addLast(new pairVO(node,-leftMinDiag));
        while(que.size()!=0){
            int size=que.size();
            while(size-->0){
                pairVO rpair = que.removeFirst();
                ans.get(rpair.vl).add(rpair.node.data);
                if(rpair.node.left!=null) que.addLast(new pairVO(rpair.node.left,rpair.vl-1));
                if(rpair.node.right!=null) que.addLast(new pairVO(rpair.node.right,rpair.vl+0));

            }
        }
        for(ArrayList<Integer>ar:ans){
            System.out.println(ar);
        }

    }


//Traversal Set Iterative

public static class tpair{
    Node node=null;
    boolean selfdone = false;
    boolean leftdone = false;
    boolean rightdone = false;

    tpair(Node node,boolean selfdone,boolean leftdone,boolean rightdone){
        this.node=node;
        this.selfdone=selfdone;
        this.leftdone=leftdone;
        this.rightdone=rightdone;
    }
}

public static void preOrderItr(Node node){
    Stack<tpair> st = new Stack<>();
    st.push(new tpair(node,false,false,false));

    while(st.size()!=0){
        if(st.peek().selfdone==false){
            st.peek().selfdone=true;
            System.out.print(st.peek().node.data + " ");
        }
        else if(st.peek().leftdone==false){
            st.peek().leftdone=true;
            if(st.peek().node.left!=null){
                st.push(new tpair(st.peek().node.left,false,false,false));
            }
        }else if(st.peek().rightdone==false){
            st.peek().rightdone=true;
            if(st.peek().node.right!=null){
                st.push(new tpair(st.peek().node.right,false,false,false));
            }
        }else{
            st.pop();
        }
    }
}

//Tree Construction

//From Pre And In Order

public static Node ConstTreePreAndIn(int []pre,int psi,int pei,int []in,int isi,int iei){
    if(psi>pei || isi>iei) return null;

    Node node = new Node(pre[psi]);
    int idx = isi;
    while(in[idx]!=pre[psi]){
        idx++;
    }
    int tnel = idx - isi;

node.left = ConstTreePreAndIn(pre, psi+1 ,psi+tnel ,in, isi , idx-1);
node.right = ConstTreePreAndIn(pre, psi+tnel+1 , pei ,in, idx+1 , iei );

return node;
}


public static void main(String[] args) {
    int []arr = {10,20,30,40,50,60,70,80,90,100,110}; //sorted array required
    Node root = createBST(arr, 0, arr.length-1);
    display(root);
    System.out.println();
    // System.out.println(findInBST(root,0));
    // System.out.println(findInBSTitr(root,0));
    //System.out.println(isBST(root));
    //kfar(30,root,2);
    //lineWiseLevelOrder(root);
    // leftView(root);
    // rightView(root);
    // verticalOrder(root);
    // verticalOrderSum(root);
    diagonalOrder(root);
}


}

File: lecture_014_Tree/treePrac.cpp
#include<iostream>
#include<vector>

#define vi vector<int>
using namespace std;


class Node{
    public:
    int data;
    Node* left;
    Node* right;

    Node(int data){
        this->data = data;
    }

    Node(int data,Node* left,Node* right){
        this->data = data;
        this->left = left;
        this->right = right;
    }
};

int idx=0;
Node* createTree(vi &arr){
    if(idx==arr.size() || arr[idx]==-1){
        idx++;
        return nullptr;
    }

    Node* node = new Node(arr[idx]);
    idx++;
    node->left = createTree(arr);
    node->right = createTree(arr);

    return node;
}

void display(Node* node){
    if(node==nullptr) return;

    string ans="";
    ans+=node->left!=nullptr?to_string(node->left->data):".";
    ans+="<- " + to_string(node->data) + " ->"; 
    ans+=node->right!=nullptr?to_string(node->right->data):".";

    cout<<ans<<endl;
    display(node->left);
    display(node->right);
}

bool findData(Node* node,int data){
    if(node==nullptr) return false;
    if(node->data == data) return true;

    bool res = false;
    res = res || findData(node->left,data);
    res = res || findData(node->right,data);

    return res;
}

int height(Node* node){
    return node==nullptr?-1:max(height(node->left),height(node->right))+1;
}

int size(Node* node){
    return node==nullptr?0:size(node->left)+size(node->right)+1;
}

int maxEle(Node* node){
    if(node==nullptr) return -1e8;
    return max(max(node->left->data,node->right->data),node->data);
}

bool nodeToRootPath(Node* node,vi &ans,int data){
    if(node==nullptr) return false;
    if(node->data==data){
        ans.push_back(data);
        return true;
    }

    bool res = false;
    res = res || nodeToRootPath(node->left,ans,data);
    res = res || nodeToRootPath(node->right,ans,data);

    if(res) ans.push_back(node->data);
    return res;
}

int LCA(Node* node,int data1,int data2){
    vi list1 ;
    vi list2;
    nodeToRootPath(node,list1,data1);
    nodeToRootPath(node,list2,data2);

    int prev;
    int i=list1.size()-1;
    int j=list2.size()-1;

    while(i>=0 && j>=0){

        if(list1[i]!=list2[j]) break;

        prev=list1[i];
        i--;
        j--;
    } 

    return prev;
}

int diameter(Node* node){
    if(node==nullptr) return 0;

    int ld =diameter(node->left);
    int rd =diameter(node->right);

    int lh=height(node->left);
    int rh=height(node->right);

    int myDia = lh+rh+2;
    return max(max(ld,rd),myDia);
}

int maxSum = -1e8; // we cant make this global for various test cases as it needs to be initialized again n again
int leafToleafMaxSum(Node* node){
    if(node->left ==nullptr && node->right==nullptr) return node->data;

    int leftnodetoleaf = leafToleafMaxSum(node->left);
    int rightnodetoleaf = leafToleafMaxSum(node->right);
    
    if(node->left !=nullptr && node->right!=nullptr){
        int sum = leftnodetoleaf + rightnodetoleaf + node->data;
        maxSum = max(maxSum,sum);
        return max(leftnodetoleaf,rightnodetoleaf) + node->data;
    }

    return node->left==nullptr?rightnodetoleaf:leftnodetoleaf + node->data;
}


//BST in java


int main(){
    vi arr ={10,20,40,60,-1,-1,70,-1,-1,50,80,
           -1,-1,-1,30,90,-1,110,150,-1,-1,-1,100,120,-1,-1,-1};
    Node* node = createTree(arr);
    display(node);
    cout<<endl;
    //cout<<height(node)<<endl;
    //cout<<size(node)<<endl;
    // vi ans;
    // nodeToRootPath(node,ans,70);
    // for(int ele: ans){
    //     cout<<ele<<" ";
    // }
    cout<<LCA(node,60,70);
return 0;
}